### PR TITLE
Speed up learning guide

### PR DESCRIPTION
**Pull request by Dantemss**
Ready for review.

### CODE REVIEW

**Comment by Dantemss** [comment]
Oh I realize that there's some potential to combine a lot of BL queries, so those can be even more optimized (I only did local optimizations, didn't have time to optimize BL calls).

**Comment by nathanstitt** [comment]
Very cool.  Loads in 2.1 seconds for me now, used to take over 40.

File: app/routines/reset_practice_widget.rb [hunk 1]
```diff
@@ -54,6 +54,7 @@ def exec(role:, exercise_source:, page_ids: [], book_part_ids: [], randomize: tr
       step = Tasks::Models::TaskStep.new(task: task)
 
       step.tasked = TaskExercise[exercise: exercise, task_step: step]
+      step.add_related_content(get_related_content_for(exercise)) unless exercise_source == :fake
```
Comment (kevinburleigh75) [inline app/routines/reset_practice_widget.rb]: Why don't we need related content for exercises from the `:fake` source?

File: app/routines/reset_practice_widget.rb [hunk 2]
```diff
@@ -69,6 +70,22 @@ def exec(role:, exercise_source:, page_ids: [], book_part_ids: [], randomize: tr
...
+    { title: page.title, chapter_section: page.chapter_section }
+  end
+
+  def content_exercise_page(content_exercise)
+    los = content_exercise.los + content_exercise.aplos
+
+    pages = Content::Models::Page.joins{page_tags.tag}
+                                 .where{page_tags.tag.value.in los}
+
+    raise "#{pages.count} pages found for exercise #{content_exercise.url}" unless pages.one?
```
Comment (kevinburleigh75) [inline app/routines/reset_practice_widget.rb]: Are we planning to change this before we allow multiple book versions to be imported?  Will it explode if we don't?

Comment (Dantemss) [inline app/routines/reset_practice_widget.rb]: Yeah it'll go kaboom.

File: app/routines/reset_practice_widget.rb (hunk 1 above)
Comment (Dantemss) [inline app/routines/reset_practice_widget.rb]: `:fake` exercises are not even persisted in the DB so they have no page... I suppose we could create a fake page, but that's weird too...

File: app/subsystems/tasks/get_tasks.rb [hunk 3]
```diff
@@ -8,16 +8,8 @@ class Tasks::GetTasks
...
-                     where{entity_role_id.in role_ids}.
-                     collect{|tasking| tasking.task}
-
-    # This was temporarily removed due to a problem (with entity_extensions?):
-    # outputs[:tasks] = Entity::Task.joins{taskings}
-    #                               .where{taskings.entity_role_id.in role_ids}
-
-    outputs[:tasks] = entity_tasks
+    outputs[:tasks] = Entity::Task.joins{taskings}
+                                  .where{taskings.entity_role_id.in role_ids}
```
Comment (kevinburleigh75) [inline app/subsystems/tasks/get_tasks.rb]: Yay!

Comment (Dantemss) [inline app/subsystems/tasks/get_tasks.rb]: Funny enough I think in the end I removed this routine from the learning guide routine.

**Pull request merged**